### PR TITLE
Avoid rescanning cli side license scans

### PR DESCRIPTION
**Pull request by spatten**
# Overview

[Delivers ANE-190](https://fossa.atlassian.net/browse/ANE-190)

In the current implementation of cli-side license scanning, we will run Themis on the vendored dependencies every time we run `fossa analyze`. This can be slow.

To make this even more painful, we then upload the results to Core, Core runs through its standard queueing logic and does not actually do anything with the data we just uploaded because the revision already exists and is resolved.

So this feature checks with Core to see if a revision exists and is resolved before scanning it. If it exists and is resolved, then we skip the scan.

The endpoint we are hitting on Core is being added in https://github.com/fossas/FOSSA/pull/7710

The CLI may be updated before the instance of Core it's pointing at has been updated to include the new endpoint, so we also added a flag in the response to `GET /api/cli/organization` that lets us know if Core supports querying for revision info. If it does not, we just scan all revisions.

## Acceptance criteria

* We should check with Core to see if a revision exists and is resolved before running a cli-side license scan on it
* If a revision exists and is resolved, we should skip the scan
* If a revision does not exist or is not yet resolved, we should run the scan
* If Core does not support the "get revision info" endpoint, then we will not check with Core to see if a revision exists and we will just rescan everything
* The results on Core should look the same as if we rescanned it
* The debug logs should let us know what actually happened: which projects were skipped and which were scanned

## Testing plan

Run a cli-side license scan on a project with multiple vendored dependencies.

The new Core API must work with a push-only API token, so use a push-only API token for all of these tests.

### Setup

We're going to scan the `two-vendored-dependencies` project from the example-projects repo: https://github.com/fossas/example-projects/tree/main/fossa-deps/two-vendored-dependencies

I've copied that into ~/fossa/license-scan-dirs/two-vendored-dependencies

### Run with current Core

Set your current core to the `develop` branch, then scan the `two-vendored-dependencies` project:

```
cabal run fossa -- analyze --debug --experimental-native-license-scan -e http://localhost:9578 ~/fossa/license-scan-dirs/two-vendored-dependencies
```

Run this multiple times. In the logs, it should let you know that it's rescanning every time:

```
[DEBUG] Skipping vendored dependency scans not supported on your FOSSA instance. All vendored dependencies will be scanned
[DEBUG] License Scanning 'first-dir-aaa11' at 'first-dir'
[DEBUG] Uploading 'first-dir-aaa11' to secure S3 bucket
[DEBUG] License Scanning 'second-dir-aaa11' at 'second-dir'
[DEBUG] Uploading 'second-dir-aaa11' to secure S3 bucket
```

### Run with Core with the new endpoint

Checkout the core PR with the new endpoint, `cli-revisions-endpoint`

Run the scan again. The debug logs should let you know that the revisions have already been scanned and are being skipped:

```
[DEBUG] All vendored dependencies have already been scanned by FOSSA. Skipping vendored dependency license scans.
```

Now change one of the vendored dependencies and scan again. The one you changed should be rescanned:

```
uuidgen > ~/fossa/license-scan-dirs/two-vendored-dependencies/second-dir/VERSION
cabal run fossa -- analyze --debug --experimental-native-license-scan -e http://localhost:9578 ~/fossa/license-scan-dirs/two-vendored-dependencies
Up to date
[DEBUG] Loading configuration file from "/Users/scott/fossa/license-scan-dirs/two-vendored-dependencies/"
[DEBUG] Some vendored dependencies have already been scanned by FOSSA. Skipping these vendored dependencies: [VendoredDependency {vendoredName = "first-dir-aaa11", vendoredPath = "first-dir", vendoredVersion = Just "ceb0678c45554cee7ca5cb0cf627a076"}]
[DEBUG] License Scanning 'second-dir-aaa11' at 'second-dir'
[DEBUG] Uploading 'second-dir-aaa11' to secure S3 bucket
```

Finally, change both of the projects and rescan. The debug logs should say that everything needs scanning.

```
uuidgen > ~/fossa/license-scan-dirs/two-vendored-dependencies/second-dir/VERSION
uuidgen > ~/fossa/license-scan-dirs/two-vendored-dependencies/first-dir/VERSION
cabal run fossa -- analyze --debug --experimental-native-license-scan -e http://localhost:9578 ~/fossa/license-scan-dirs/two-vendored-dependencies
Up to date
[DEBUG] Loading configuration file from "/Users/scott/fossa/license-scan-dirs/two-vendored-dependencies/"
[DEBUG] All vendored dependencies are un-scanned and require license scanning
[DEBUG] License Scanning 'first-dir-aaa11' at 'first-dir'
[DEBUG] Uploading 'first-dir-aaa11' to secure S3 bucket
[DEBUG] License Scanning 'second-dir-aaa11' at 'second-dir'
[DEBUG] Uploading 'second-dir-aaa11' to secure S3 bucket
```

Check on Core to make sure that the new revision you just added is scanned and shown

Screenshot 2022-05-19 at 11 42 34 AM

## Checklist

- [X] I added tests for this PR's change (or explained in the PR description why tests don't make sense).
- [x] If this change is externally visible, I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.

### CODE REVIEW

File: test/Test/Fixtures.hs [hunk 1]
```diff
@@ -206,3 +226,141 @@ issuesPending =
...
+    "second-archive-test"
+    "0.0.1"
+
+archives :: [Archive]
+archives = [firstArchive, secondArchive]
+
+locators :: NonEmpty Locator
+locators = firstLocator :| [secondLocator]
+
+firstLicenseSourceUnit :: LicenseSourceUnit
```
Comment (spatten) [inline test/Test/Fixtures.hs:295]: `firstLicenseSourceUnit` and `secondLicenseSourceUnit` will become much smaller once we can ignore the request sent to a test

**Comment by spatten** [comment]
The Windows-build tests are failing right now for a known reason that Drew and I are working on fixing.

This is ready to review ignoring that, though.

File: src/App/Fossa/LicenseScanner.hs [hunk 2]
```diff
@@ -303,31 +308,89 @@ licenseScanSourceUnit ::
   ( Has Diagnostics sig m
   , Has (Lift IO) sig m
   , Has StickyLogger sig m
+  , Has Logger sig m
   , Has Exec sig m
   , Has ReadFS sig m
   , Has FossaApiClient sig m
   ) =>
+  Bool ->
```
Comment (kitified) [inline src/App/Fossa/LicenseScanner.hs:316]: Can we make this an enum? Bools as arguments feel very unclear, especially for public functions in other modules.
Suggestion:
```haskell
data LicenseScanMode
  = LicenseScanModeSkipPreviouslyScanned
  | LicenseScanModeFull
```
Or similar (and similarly refactor the API calls and any call sites).

File: src/App/Fossa/LicenseScanner.hs [hunk 3]
```diff
@@ -303,31 +308,89 @@ licenseScanSourceUnit ::
...
+  (needScanningDeps, skippedDeps) <-
+    if skippingSupported
+      then findDepsThatNeedScanning uniqDepsWithVersions orgId
+      else pure (NE.toList uniqDepsWithVersions, [])
+
+  -- Debug logs giving info about which vendored deps were actually scanned
+  -- If none of the dependencies need scanning, then log that, but we still need to do `finalizeLicenseScan` so keep going
+  _ <- case (needScanningDeps, skippingSupported) of
+    (_, False) ->
+      logDebug "Skipping vendored dependency scans not supported on your FOSSA instance. All vendored dependencies will be scanned"
```
Comment (kitified) [inline src/App/Fossa/LicenseScanner.hs:339]: [optional] I think referring to this as "skipping" is potentially confusing. Can we reword this in a more user friendly way?
Something like (but feel free to editorialize):
```haskell
logDebug "This version of the FOSSA service does not support enumerating previously scanned vendored dependencies."
logDebug "Performing a full scan of all vendored dependencies even if they have been scanned previously."
```

Some comments below are also rewordings, please also treat them as optional and/or modify as you see fit.

File: src/App/Fossa/LicenseScanner.hs [hunk 4]
```diff
@@ -303,31 +308,89 @@ licenseScanSourceUnit ::
...
+    ([], True) -> do
+      logDebug "All vendored dependencies have already been scanned by FOSSA. Skipping vendored dependency license scans."
+    (_, True) -> do
+      case skippedDeps of
+        [] -> logDebug "All vendored dependencies are un-scanned and require license scanning"
```
Comment (kitified) [inline src/App/Fossa/LicenseScanner.hs:344]:
```suggestion
        [] -> logDebug "None of the current vendored dependencies have been previously scanned."
```

File: src/App/Fossa/LicenseScanner.hs [hunk 5]
```diff
@@ -303,31 +308,89 @@ licenseScanSourceUnit ::
...
+      then findDepsThatNeedScanning uniqDepsWithVersions orgId
+      else pure (NE.toList uniqDepsWithVersions, [])
+
+  -- Debug logs giving info about which vendored deps were actually scanned
+  -- If none of the dependencies need scanning, then log that, but we still need to do `finalizeLicenseScan` so keep going
+  _ <- case (needScanningDeps, skippingSupported) of
+    (_, False) ->
+      logDebug "Skipping vendored dependency scans not supported on your FOSSA instance. All vendored dependencies will be scanned"
+    ([], True) -> do
+      logDebug "All vendored dependencies have already been scanned by FOSSA. Skipping vendored dependency license scans."
```
Comment (kitified) [inline src/App/Fossa/LicenseScanner.hs:341]:
```suggestion
      logDebug "All of the current vendored dependencies have been previously scanned, reusing previous results."
```

File: src/App/Fossa/LicenseScanner.hs [hunk 6]
```diff
@@ -303,31 +308,89 @@ licenseScanSourceUnit ::
...
+    (_, True) -> do
+      case skippedDeps of
+        [] -> logDebug "All vendored dependencies are un-scanned and require license scanning"
+        _ -> logDebug . pretty $ "Some vendored dependencies have already been scanned by FOSSA. Skipping these vendored dependencies: " <> show skippedDeps
```
Comment (kitified) [inline src/App/Fossa/LicenseScanner.hs:345]:
```suggestion
        _ -> do
          logDebug "Some of the current vendored dependencies have already been scanned by FOSSA."
          logDebug . pretty $ "Reusing previous results for the following vendored dependencies: " <> show skippedDeps
```

File: src/App/Fossa/LicenseScanner.hs [hunk 7]
```diff
@@ -303,31 +308,89 @@ licenseScanSourceUnit ::
...
+
+shouldScanRevision :: [RevisionInfo] -> OrgId -> VendoredDependency -> Bool
+shouldScanRevision revisionInfos orgId vdep = all (locatorMatches vdep orgId) revisionInfos
+  where
+    locatorMatches :: VendoredDependency -> OrgId -> RevisionInfo -> Bool
+    locatorMatches VendoredDependency{..} org RevisionInfo{..} =
+      locatorUrl /= revisionInfoLocator || not revisionInfoResolved
+      where
+        locator = Locator{locatorFetcher = "archive", locatorProject = vendoredName, locatorRevision = vendoredVersion}
+        locatorUrl = renderLocatorUrl org locator
```
Comment (kitified) [inline src/App/Fossa/LicenseScanner.hs:396]: If I'm reading this correctly, we will not rescan a vendored dependency if:
- The version in `fossa-deps.yaml` has not changed, and
- A file on disk inside the directory has changed (I think we're only hashing if there is no version provided).

If I'm correct, I think that's a little unintuitive, but I'm okay with it if that's intentional. Will you please update the documentation and add a doc comment to this function making it clear this is intended functionality?

File: src/Fossa/API/Types.hs [hunk 8]
```diff
@@ -402,3 +405,14 @@ useApiOpts opts = case useURI serverURI of
...
+data RevisionInfo = RevisionInfo
+  { revisionInfoLocator :: Text
+  , revisionInfoResolved :: Bool
+  }
+  deriving (Eq, Ord, Show)
+
+instance FromJSON RevisionInfo where
+  parseJSON = withObject "RevisionInfo" $ \obj ->
+    RevisionInfo <$> obj .: "locator"
+      <*> obj .: "resolved"
```
Comment (kitified) [inline src/Fossa/API/Types.hs:418]: [nit, optional] I think we generally put all these operators on newlines if we do any of them:
```suggestion
    RevisionInfo
      <$> obj .: "locator"
      <*> obj .: "resolved"
```

**Review by kitified — APPROVED** [review]
lgtm after comments - all optional

File: src/App/Fossa/LicenseScanner.hs (hunk 7 above)
Comment (spatten) [reply src/App/Fossa/LicenseScanner.hs]: You're right, this is not at all intuitive.

This logic was adopted from how archive uploads work.

The reasoning is that even if we did run a license scan on the dependency, an already analyzed revision exists on core already, so the build would not get queued and the data we send up would just be thrown away.

I'll add a comment to explain this

File: src/App/Fossa/ManualDeps.hs [hunk 9]
```diff
@@ -165,21 +166,22 @@ scanAndUpload ::
...
+        let doNative = orgDoLocalLicenseScan org
         if doNative
           then pure CLILicenseScan
           else -- If they've selected native scanning, but the server doesn't support it,
           -- we should let them know.
           -- TODO: Add a --forbid-archive-upload CLI flag
             logWarn "Server does not support native license scanning" $> ArchiveUpload
       else pure ArchiveUpload
-
+  let vendoredDependencyScanMode = if (orgSupportsAnalyzedRevisionsQuery org) then VendoredDependencyScanModeSkipPreviouslyScanned else VendoredDependencyScanModeSkippingNotSupported
```
Comment (scruffystuffs) [inline src/App/Fossa/ManualDeps.hs:181]: [fmt]: I am legitimately shocked that the formatter did not break this line up automatically.  Please break this line up.

File: src/App/Fossa/VendoredDependency.hs [hunk 10]
```diff
@@ -47,6 +48,10 @@ instance FromJSON VendoredDependency where
       <*> obj .: "path"
       <*> (unTextLike <$$> obj .:? "version")
       <* forbidMembers "vendored dependencies" ["type", "license", "url", "description"] obj
+data VendoredDependencyScanMode
+  = VendoredDependencyScanModeSkipPreviouslyScanned
+  | VendoredDependencyScanModeSkippingNotSupported
+  deriving (Eq, Show)
```
Comment (scruffystuffs) [inline src/App/Fossa/VendoredDependency.hs:54]: [nit]: Please add `Ord` to the list of derives

File: src/Control/Carrier/FossaApiClient/Internal/FossaAPIV1.hs [hunk 11]
```diff
@@ -366,6 +380,32 @@ getProject apiopts ProjectRevision{..} = fossaReq $ do
...
+  let locatorBody = constructLocatorBody orgId vDeps
+  responseBody <$> req POST (getAnalyzedRevisionsEndpoint baseUrl) (ReqBodyJson locatorBody) jsonResponse baseOpts
+  where
+    constructLocatorBody :: OrgId -> NonEmpty VendoredDependency -> GetAnalyzedRevisionsBody
+    constructLocatorBody orgId = GetAnalyzedRevisionsBody <$> NE.map (constructLocator orgId)
+      where
+        constructLocator :: OrgId -> VendoredDependency -> Text
+        constructLocator org VendoredDependency{..} = do
+          let locator = Locator{locatorFetcher = "archive", locatorProject = vendoredName, locatorRevision = vendoredVersion}
+          renderLocatorUrl org locator
```
Comment (scruffystuffs) [inline src/Control/Carrier/FossaApiClient/Internal/FossaAPIV1.hs:405]: [required]: This function should probably be cleaned up, it's a bit messy:
- We're using `where` clauses inside of `do` notation, we should either use `let` bindings or separate functions.
- The `where` clauses are nested, with further nested inner `do` notation and `let` bindings.  I think this function can get a lot flatter.

File: src/Control/Effect/FossaApiClient.hs [hunk 12]
```diff
@@ -86,6 +89,7 @@ data FossaApiClientF a where
   GetLatestScan :: Locator -> ProjectRevision -> FossaApiClientF ScanResponse
   GetOrganization :: FossaApiClientF Organization
   GetProject :: ProjectRevision -> FossaApiClientF Project
+  GetAnalyzedRevisions :: (NonEmpty VendoredDependency) -> FossaApiClientF [Text]
```
Comment (scruffystuffs) [inline src/Control/Effect/FossaApiClient.hs:92]: [nit]: Kill parens
```suggestion
  GetAnalyzedRevisions :: NonEmpty VendoredDependency -> FossaApiClientF [Text]
```

File: src/Control/Effect/FossaApiClient.hs [hunk 13]
```diff
@@ -171,6 +175,9 @@ assertRevisionBinaries locator fprints = sendSimple (AssertRevisionBinaries loca
 assertUserDefinedBinaries :: Has FossaApiClient sig m => IAT.UserDefinedAssertionMeta -> [Fingerprint Raw] -> m ()
 assertUserDefinedBinaries meta fprints = sendSimple (AssertUserDefinedBinaries meta fprints)
 
+getAnalyzedRevisions :: Has FossaApiClient sig m => (NonEmpty VendoredDependency) -> m ([Text])
```
Comment (scruffystuffs) [inline src/Control/Effect/FossaApiClient.hs:178]: [nit]: Kill parens
```suggestion
getAnalyzedRevisions :: Has FossaApiClient sig m => NonEmpty VendoredDependency -> m [Text]
```

File: test/App/Fossa/LicenseScanner/LicenseScannerSpec.hs [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+    it' "should scan all if the revisions are still being scanned" $ do
+      expectGetApiOpts
+      expectGetOrganization
+      expectAllScansInProgress
+      expectGetSignedUrl PackageRevision{packageName = "first-archive-test", packageVersion = "0.0.1"}
+      expectUploadLicenseScanResult Fixtures.firstLicenseSourceUnit
+      expectGetSignedUrl PackageRevision{packageName = "second-archive-test", packageVersion = "0.0.1"}
+      expectUploadLicenseScanResult Fixtures.secondLicenseSourceUnit
+      expectFinalizeScan Fixtures.archives
+      locators <- runStack . ignoreLogger . runDiagnostics . ignoreStickyLogger . runExecIO . runReadFSIO $ licenseScanSourceUnit VendoredDependencyScanModeSkipPreviouslyScanned scanDir Fixtures.vendoredDeps
```
Comment (scruffystuffs) [inline test/App/Fossa/LicenseScanner/LicenseScannerSpec.hs:131]: [required]: remove effect handlers

File: test/App/Fossa/LicenseScanner/LicenseScannerSpec.hs [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+          ls `shouldBe'` Fixtures.locators
+
+    it' "should scan one if one revision is still being scanned" $ do
+      expectGetApiOpts
+      expectGetOrganization
+      expectOneScanInProgress
+      expectGetSignedUrl PackageRevision{packageName = "first-archive-test", packageVersion = "0.0.1"}
+      expectUploadLicenseScanResult Fixtures.firstLicenseSourceUnit
+      expectFinalizeScan Fixtures.archives
+      locators <- runStack . ignoreLogger . runDiagnostics . ignoreStickyLogger . runExecIO . runReadFSIO $ licenseScanSourceUnit VendoredDependencyScanModeSkipPreviouslyScanned scanDir Fixtures.vendoredDeps
```
Comment (scruffystuffs) [inline test/App/Fossa/LicenseScanner/LicenseScannerSpec.hs:144]: [required]: remove effect handlers

File: test/App/Fossa/LicenseScanner/LicenseScannerSpec.hs [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+
+    it' "should always scan all if vendor dependency skipping is not supported" $ do
+      expectGetApiOpts
+      expectGetOrganization
+      expectGetSignedUrl PackageRevision{packageName = "first-archive-test", packageVersion = "0.0.1"}
+      expectUploadLicenseScanResult Fixtures.firstLicenseSourceUnit
+      expectGetSignedUrl PackageRevision{packageName = "second-archive-test", packageVersion = "0.0.1"}
+      expectUploadLicenseScanResult Fixtures.secondLicenseSourceUnit
+      expectFinalizeScan Fixtures.archives
+      locators <- runStack . ignoreLogger . runDiagnostics . ignoreStickyLogger . runExecIO . runReadFSIO $ licenseScanSourceUnit VendoredDependencyScanModeSkippingNotSupported scanDir Fixtures.vendoredDeps
```
Comment (scruffystuffs) [inline test/App/Fossa/LicenseScanner/LicenseScannerSpec.hs:158]: [required]: remove effect handlers

**Review by scruffystuffs — CHANGES_REQUESTED** [review]
Left quite a few comments, but I'd like to block on the `[required]:` comments at the very least.

File: test/App/Fossa/LicenseScannerSpec.hs [hunk 17]
```diff
@@ -1,15 +1,30 @@
+{-# LANGUAGE QuasiQuotes #-}
```
Comment (scruffystuffs) [inline test/App/Fossa/LicenseScannerSpec.hs:1]: I think this extension can be removed.
```suggestion
```

**Review by scruffystuffs — APPROVED** [review]
Looks good, only nits left.

File: test/App/Fossa/LicenseScannerSpec.hs [hunk 18]
```diff
@@ -42,25 +61,124 @@ expectedCombinedUnit =
...
-spec =
-  -- this test only exists to prevent merging the commented out analyzers
+spec = do
   describe "combineLicenseUnits" $ do
     it "should combine two MIT units" $
       combineLicenseUnits [unitOne, unitTwo] `shouldBe` [expectedCombinedUnit]
     it "should not combine two units with different licenses" $
       combineLicenseUnits [unitOne, unitTwo{licenseUnitName = "AGPL"}] `shouldBe` [unitTwo{licenseUnitName = "AGPL"}, unitOne]
+
+  describe "licenseScanSourceUnits" $ do
```
[the remark this answers is not in the document]
Comment (spatten) [reply test/App/Fossa/LicenseScannerSpec.hs:90]: Yeah, I'm loving working with Drew's MockApi

**Pull request merged**